### PR TITLE
[rllib] Implement custom metrics

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

This adds support for custom callbacks that are run during policy evaluation. We pass in a reference to the stateful `episode` object, which can be used to store arbitrary state and emit custom metrics.

Right now the only aggregation type for custom metrics is `mean`, but we could potentially add more in the future. For more complex types of custom metrics, you could write those separately in the callbacks.

Custom metrics are included in the global logging. From the cartpole example:
screenshot from 2018-10-26 15-12-44

TODO:
- [ ] Add docs
- [ ] Add tests

## Related issue number

https://github.com/ray-project/ray/issues/3122

### CODE REVIEW

**Review by Nas2586 — APPROVED** [review]

File: python/ray/rllib/examples/custom_metrics_and_callbacks.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+"""
+
+import argparse
+import numpy as np
+
+import ray
+from ray import tune
+
+
+def on_episode_start(info):
```
Comment (richardliaw) [inline python/ray/rllib/examples/custom_metrics_and_callbacks.py]: One proposal for this - https://keras.io/callbacks/#create-a-callback

**Comment by ericl** [comment]
Yeah that's not really needed since you can store state in the episode
object.

**Comment by ericl** [comment]
@richardliaw ready for review

File: doc/source/rllib-training.rst [hunk 2]
```diff
@@ -224,6 +224,78 @@ Sometimes, it is necessary to coordinate between pieces of code that live in dif
...
+
+    python ray/python/ray/rllib/train.py --env=PongDeterministic-v4 \
+        --run=A2C --config '{"num_workers": 2, "log_level": "DEBUG"}'
+
+Callbacks and Custom Metrics
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can provide callback functions to be called at points during policy evaluation. These functions have access to an info dict containing state for the current episode. Custom state can be stored for the episode in the ``info["episode"].user_data`` dict, and custom scalar metrics reported by saving values to the ``info["episode"].custom_metrics`` dict. These custom metrics will be averaged and reported as part of training results. The following example (full code `here <https://github.com/ray-project/ray/blob/master/python/ray/rllib/examples/custom_metrics_and_callbacks.py>`__) logs a custom metric from the environment:
+
+.. code-block:: python
```
Comment (richardliaw) [inline doc/source/rllib-training.rst]: Looking at the example and this doc, I wouldn't know the contents of the info dict for all of the methods, can you document?

**Review by richardliaw — COMMENTED** [review]
comments on documentation...

**Comment by ericl** [comment]
All relevant tests passing.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**